controller/volume_rebuilding_controller.go

```diff
@@ -192,12 +279,134 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
 	}()
 
 	rebuildingAttachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeRebuildingController, volName)
+	deleteVATicketRequired := true
+	defer func() {
+		if deleteVATicketRequired {
+			delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+		}
+	}()
 
-	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+	if !vol.DeletionTimestamp.IsZero() {
+		vbc.logger.Infof("Volume %v is deleting, skip offline rebuilding", volName)
+		return nil
+	}
+	if vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		vbc.logger.Warnf("Volume %v is faulted, skip offline rebuilding", volName)
+		return nil
+	}
 
+	isOfflineRebuildEnabled, err := vbc.isVolumeOfflineRebuildEnabled(vol.Spec.OfflineRebuilding)
+	if err != nil {
+		return err
+	}
+	if isOfflineRebuildEnabled && !util.IsHigherPriorityVATicketExisting(va, longhorn.AttacherTypeVolumeRebuildingController) {
+		if vol.Status.State == longhorn.VolumeStateDetached {
+			_, err = vbc.syncLHVolumeAttachmentForOfflineRebuild(vol, va, rebuildingAttachmentTicketID)
+			if err != nil {
+				return err
+			}
+			deleteVATicketRequired = false
+			return nil
+		}
+		if vol.Status.State == longhorn.VolumeStateAttaching {
+			deleteVATicketRequired = false
+			return nil
+		}
+
+		engine, err := vbc.getVolumeEngine(vol)
+		if err != nil {
+			return err
+		}
+		if engine == nil {
+			vbc.logger.Warnf("Volume %v engine not found, skip offline rebuilding", volName)
+			return nil
+		}
+
+		deleteVATicketRequired = false
+		if engine.Status.ReplicaModeMap == nil {
+			// wait for engine status synced
+			return nil
+		}
+		if vbc.isVolumeReplicasRebuilding(vol, engine) {
+			return nil
+		}
+	}
+
+	deleteVATicketRequired = true
```

Comment:
Is it redundant?

Reply:
No, because theline 325, we set the flag to false.